File: python/problems/leetcode/0048_rotate_image/solution.py
"""LeetCode 48: Rotate Image."""

from __future__ import annotations


class Solution:
    def rotate(self, matrix: list[list[int]]) -> None:
        n = len(matrix)
        for i in range(n):
            for j in range(i + 1, n):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
        for row in matrix:
            row.reverse()


if __name__ == "__main__":
    a = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    Solution().rotate(a)
    assert a == [[7, 4, 1], [8, 5, 2], [9, 6, 3]]
    b = [[5, 1, 9, 11], [2, 4, 8, 10], [13, 3, 6, 7], [15, 14, 12, 16]]
    Solution().rotate(b)
    assert b == [
        [15, 13, 2, 5],
        [14, 3, 4, 1],
        [12, 6, 8, 9],
        [16, 7, 10, 11],
    ]
    print("0048 OK")

File: cpp/problems/leetcode/0048_rotate_image/solution.cpp
// LeetCode 48
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int n = (int)matrix.size();
        for (int i = 0; i < n; ++i)
            for (int j = i + 1; j < n; ++j)
                swap(matrix[i][j], matrix[j][i]);
        for (auto& row : matrix) reverse(row.begin(), row.end());
    }
};

int main() {
    vector<vector<int>> a{{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
    Solution().rotate(a);
    assert((a == vector<vector<int>>{{7, 4, 1}, {8, 5, 2}, {9, 6, 3}}));
    cout << "0048 OK" << endl;
    return 0;
}
